EventListener/TagSubscriber.php

```diff
@@ -86,8 +113,13 @@ public function onKernelResponse(FilterResponseEvent $event)
         }
 
         if ($request->isMethodSafe()) {
-            // For safe requests (GET and HEAD), set cache tags on response
-            $this->cacheManager->tagResponse($response, $tags);
+            if ($event->isMasterRequest()) {
+                // For safe requests (GET and HEAD), set cache tags on response
+                $this->cacheManager->tagResponse($response, $tags);
+            } else {
+                // Retain tags to attach to response to master request
+                $this->tags = $tags;
+            }
```

Comment:
and when using ESI, each master request has its own cache, so you will want to tag the ESI requests on themselves